File: latchkey/services/linear.py
import uuid

from playwright.sync_api import BrowserContext
from playwright.sync_api import Response
from pydantic import PrivateAttr

from latchkey import curl
from latchkey.api_credentials import ApiCredentialStatus
from latchkey.api_credentials import ApiCredentials
from latchkey.api_credentials import AuthorizationBearer
from latchkey.services.base import BrowserFollowupServiceSession
from latchkey.services.base import LoginFailedError
from latchkey.services.base import Service
from latchkey.services.playwright_utils import type_like_human

DEFAULT_TIMEOUT_MS = 8000

# URL for creating a new personal API key (also used as login URL)
LINEAR_NEW_API_KEY_URL = "https://linear.app/imbue/settings/account/security/api-keys/new"


class LinearServiceSession(BrowserFollowupServiceSession):
    _is_logged_in: bool = PrivateAttr(default=False)

    def on_response(self, response: Response) -> None:
        if self._is_logged_in:
            return

        request = response.request
        url = request.url
        if not url.startswith("https://linear.app/"):
            return

        # Detect login by checking for authenticated API requests
        headers = request.all_headers()
        # Linear sets authentication cookies after successful login
        cookies = headers.get("cookie", "")
        if "linear-session" in cookies or "linear_session" in cookies:
            self._is_logged_in = True

    def _is_headful_login_complete(self) -> bool:
        return self._is_logged_in

    def _perform_browser_followup(self, context: BrowserContext) -> ApiCredentials | None:
        page = context.new_page()

        page.goto(LINEAR_NEW_API_KEY_URL)

        # Fill in the key name
        key_name = f"Latchkey-{uuid.uuid4().hex[:8]}"
        key_name_input = page.get_by_role("textbox", name="Key name")
        key_name_input.wait_for(timeout=DEFAULT_TIMEOUT_MS)
        type_like_human(page, key_name_input, key_name)

        # Click the Create button
        create_button = page.get_by_role("button", name="Create")
        create_button.wait_for(timeout=DEFAULT_TIMEOUT_MS)
        create_button.click()

        # Wait for and extract the token from span element containing lin_api_ prefix
        token_element = page.locator("span:text-matches('^lin_api_')")
        token_element.wait_for(timeout=DEFAULT_TIMEOUT_MS)

        token = token_element.text_content()
        if token is None or token == "":
            raise LoginFailedError("Failed to extract token from Linear.")

        page.close()

        return AuthorizationBearer(token=token)


class Linear(Service):
    name: str = "linear"
    base_api_urls: tuple[str, ...] = ("https://api.linear.app/",)
    login_url: str = LINEAR_NEW_API_KEY_URL

    def get_session(self) -> LinearServiceSession:
        return LinearServiceSession(service=self)

    @property
    def credential_check_curl_arguments(self) -> tuple[str, ...]:
        return (
            "-X",
            "POST",
            "-H",
            "Content-Type: application/json",
            "-d",
            '{"query": "{ viewer { id } }"}',
            "https://api.linear.app/graphql",
        )

    def check_api_credentials(self, api_credentials: ApiCredentials) -> ApiCredentialStatus:
        if not isinstance(api_credentials, AuthorizationBearer):
            return ApiCredentialStatus.INVALID

        # Linear uses GraphQL API - check credentials with a simple query
        result = curl.run_captured(
            [
                "-s",
                "-o",
                "/dev/null",
                "-w",
                "%{http_code}",
                *api_credentials.as_curl_arguments(),
                *self.credential_check_curl_arguments,
            ],
            timeout=10,
        )

        if result.stdout == "200":
            return ApiCredentialStatus.VALID
        return ApiCredentialStatus.INVALID


LINEAR = Linear()

File: src/services/linear.ts
/**
 * Linear service implementation.
 */

import type { Response, BrowserContext, Page } from 'playwright';
import { ApiCredentials, AuthorizationBare } from '../apiCredentials/base.js';
import { typeLikeHuman } from '../playwrightUtils.js';
import { Service, BrowserFollowupServiceSession, LoginFailedError } from './core/base.js';
import { fetchAccountFromEndpoint, tryParseJson } from '../apiCredentials/account.js';

const DEFAULT_TIMEOUT_MS = 8000;

const LINEAR_LOGIN_URL = 'https://linear.app/login';

// How long to wait for the post-login redirects to settle on a workspace URL.
const WORKSPACE_URL_TIMEOUT_MS = 30000;
// How long a workspace URL has to stay unchanged before it counts as final.
const WORKSPACE_URL_STABLE_MS = 1000;
const WORKSPACE_URL_POLL_INTERVAL_MS = 200;

// First path segments of linear.app URLs that are not workspace names.
const NON_WORKSPACE_PATH_SEGMENTS = new Set([
  'login',
  'logout',
  'auth',
  'oauth',
  'signup',
  'join',
  'invite',
  'magic',
]);

// GraphQL response keys that Linear also returns while the user is still
// anonymous or in the middle of logging in; seeing only these proves nothing.
const ANONYMOUS_GRAPHQL_RESPONSE_KEYS = new Set([
  'organizationMeta',
  'trackAnonymousEvent',
  'emailUserAccountAuthChallenge',
  'passkeyLoginStart',
]);

function parseWorkspaceName(url: string): string | null {
  const match = /^https:\/\/linear\.app\/([^/?#]+)/.exec(url);
  if (match === null) {
    return null;
  }
  const workspaceName = match[1] ?? '';
  return workspaceName === '' || NON_WORKSPACE_PATH_SEGMENTS.has(workspaceName)
    ? null
    : workspaceName;
}

function newApiKeyUrl(workspaceName: string): string {
  return `https://linear.app/${workspaceName}/settings/account/security/api-keys/new`;
}

/**
 * After login Linear redirects to a workspace page whose path is not stable
 * over time (e.g. `/agent`, `/inbox`), so instead of assuming a landing page we
 * wait for the redirect chain to settle on any workspace URL and take the
 * workspace name from it.
 */
async function waitForWorkspaceName(page: Page): Promise<string> {
  const deadline = Date.now() + WORKSPACE_URL_TIMEOUT_MS;
  let candidateName: string | null = null;
  let candidateSince = 0;

  while (Date.now() < deadline) {
    const workspaceName = parseWorkspaceName(page.url());
    if (workspaceName !== candidateName) {
      candidateName = workspaceName;
      candidateSince = Date.now();
    } else if (candidateName !== null && Date.now() - candidateSince >= WORKSPACE_URL_STABLE_MS) {
      return candidateName;
    }
    await page.waitForTimeout(WORKSPACE_URL_POLL_INTERVAL_MS);
  }

  throw new LoginFailedError(
    `Timed out waiting for a Linear workspace URL after login (last URL: ${page.url()}).`
  );
}

class LinearServiceSession extends BrowserFollowupServiceSession {
  private isLoggedIn = false;

  onResponse(response: Response): void {
    if (this.isLoggedIn) {
      return;
    }

    const request = response.request();
    // Empirically, when not logged in, the response data only contains keys from
    // ANONYMOUS_GRAPHQL_RESPONSE_KEYS. Otherwise it can contain many different things.
    if (request.url() === 'https://client-api.linear.app/graphql' && request.method() === 'POST') {
      if (response.status() === 200) {
        try {
          // Note: response.json() returns a Promise in Playwright
          response
            .json()
            .then((jsonData: unknown) => {
              const data = (jsonData as { data?: Record<string, unknown> }).data ?? {};
              if (Object.keys(data).some((key) => !ANONYMOUS_GRAPHQL_RESPONSE_KEYS.has(key))) {
                this.isLoggedIn = true;
              }
            })
            .catch(() => {
              // Ignore JSON parse errors
            });
        } catch {
          // Ignore errors
        }
      }
    }
  }

  protected isLoginComplete(): boolean {
    return this.isLoggedIn;
  }

  protected async performBrowserFollowup(
    context: BrowserContext,
    _oldCredentials?: ApiCredentials
  ): Promise<ApiCredentials | null> {
    const page = context.pages()[0];
    if (!page) {
      throw new LoginFailedError('No page available in browser context.');
    }

    const workspaceName = await waitForWorkspaceName(page);
    await page.goto(newApiKeyUrl(workspaceName));

    // Fill in the key name
    const keyName = this.generateAppName();
    const keyNameInput = page.locator('//*[@id="label"]');
    await keyNameInput.waitFor({ timeout: DEFAULT_TIMEOUT_MS });
    await typeLikeHuman(page, keyNameInput, keyName);

    // Click the Create button
    const createButton = page.locator('button[type="submit"]');
    await createButton.waitFor({ timeout: DEFAULT_TIMEOUT_MS });
    await createButton.click();

    // Wait for and extract the token from span element containing lin_api_ prefix
    const tokenElement = page.locator("span:text-matches('^lin_api_')");
    await tokenElement.waitFor({ timeout: DEFAULT_TIMEOUT_MS });

    const token = await tokenElement.textContent();
    if (token === null || token === '') {
      throw new LoginFailedError('Failed to extract token from Linear.');
    }

    await page.close();

    return new AuthorizationBare(token);
  }
}

export class Linear extends Service {
  readonly name = 'linear';
  readonly displayName = 'Linear';
  readonly baseApiUrls = ['https://api.linear.app/'] as const;
  readonly loginUrl = LINEAR_LOGIN_URL;
  readonly info = 'https://linear.app/developers/graphql';

  readonly credentialCheckCurlArguments = [
    '-X',
    'POST',
    '-H',
    'Content-Type: application/json',
    '-d',
    '{"query": "{ viewer { id email } }"}',
    'https://api.linear.app/graphql',
  ] as const;

  setCredentialsExample(serviceName: string): string {
    return `latchkey auth set ${serviceName} -H "Authorization: <token>"`;
  }

  override getAccount(apiCredentials: ApiCredentials): Promise<string | null> {
    return fetchAccountFromEndpoint(
      apiCredentials,
      this.credentialCheckCurlArguments,
      (responseBody) => {
        const data = tryParseJson(responseBody) as {
          data?: { viewer?: { email?: string; id?: string } };
        } | null;
        return data?.data?.viewer?.email ?? data?.data?.viewer?.id ?? null;
      }
    );
  }

  override getSession(appNamePrefix: string): LinearServiceSession {
    return new LinearServiceSession(this, appNamePrefix);
  }
}

export const LINEAR = new Linear();
